File: java/src/strategy/implementations/Knuth2.java
package strategy.implementations;

import strategy.Answer;
import strategy.Code;

/**
 * In contrast to Knuth this strategy does *not* try to minimize the travel
 * distance.
 */
public class Knuth2 extends BasicStrategy {

	public Knuth2(int codeLength) {
		super(codeLength);
	}

	@Override
	public Code guess(Answer answer) {
		this.removeInconsistentCodes(this.consistentCodes, this.lastGuess,
				answer);
		this.lastGuess = consistentCodes.get(0);
		int maxMinimum = 0;
		for (Code code : this.allCodes) {
			int minimum = Integer.MAX_VALUE;
			for (Answer a : this.allAnswers) {
				int removedCodesSize = getConsistentCodes(
						this.consistentCodes, code, a).size();
				minimum = Math.min(removedCodesSize, minimum);
			}
			if (minimum > maxMinimum) {
				maxMinimum = minimum;
				this.lastGuess = code;
			}
		}
		this.lastButton = this.lastGuess.get(this.CODE_LENGTH - 1);
		return this.lastGuess;
	}

	@Override
	public String toString() {
		return "Knuth";
	}

}

File: lua/strategies/knuth2.lua
-- In contrast to knuth this strategy does *not* try to minimize the travel
-- distance.

local util = require 'strategies.util'
local createAllAnswers = util.createAllAnswers
local createAllCodes = util.createAllCodes
local getConsistentCodes = util.getConsistentCodes
local getShortestCode = util.getShortestCode
local min = math.min

-- Main logic

local allCodes = createAllCodes(CODE_LENGTH)
local allAnswers = createAllAnswers(CODE_LENGTH)
local consistentCodes = {}
local lastGuess = nil
local lastButton = {x=0,y=0}

local function reset()
  for i,code in ipairs(allCodes) do
    consistentCodes[i] = code
  end
  lastGuess = getShortestCode(lastButton, consistentCodes)
  lastButton = lastGuess[CODE_LENGTH]
  return lastGuess
end

local function guess(blacks, whites)
  consistentCodes = getConsistentCodes(consistentCodes, lastGuess, 
      blacks, whites)
  lastGuess = consistentCodes[1]
  local totalSize = #consistentCodes
  local maxScore = 1
  for _,code in ipairs(allCodes) do
    local score = math.huge
    for _,answer in ipairs(allAnswers) do
      local removals = totalSize - #getConsistentCodes(consistentCodes, code,
          answer.blacks, answer.whites)
      score = min(removals, score)
    end
    if score > maxScore then
      maxScore = score
      lastGuess = code
    end
  end
  lastButton = lastGuess[CODE_LENGTH]
  return lastGuess
end

return {
  reset = reset,
  guess = guess
}
